lib/web_ui/lib/src/engine/html/shader_mask.dart

```diff
@@ -0,0 +1,376 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.10
+part of engine;
+
+/// A surface that applies a [Shader] to its children.
+class PersistedShaderMask extends PersistedContainerSurface
+    implements ui.ShaderMaskEngineLayer {
+  PersistedShaderMask(
+    PersistedShaderMask? oldLayer,
+    this.shader,
+    this.maskRect,
+    this.blendMode,
+  ) : super(oldLayer);
+
+  @override
+  html.Element? get childContainer => _childContainer;
+
+  /// The dedicated child container element that's separate from the
+  /// [rootElement] is used to compensate for the coordinate system shift
+  /// introduced by the [rootElement] translation.
+  html.Element? _childContainer;
+
+  final ui.Shader shader;
+  final ui.Rect maskRect;
+  final ui.BlendMode blendMode;
+  html.Element? _shaderElement;
+  bool containerVisible = true;
+
+  @override
+  void adoptElements(PersistedShaderMask oldSurface) {
+    super.adoptElements(oldSurface);
+    _childContainer = oldSurface._childContainer;
+    _shaderElement = oldSurface._shaderElement;
+    oldSurface._childContainer = null;
+    oldSurface._shaderElement = null;
+  }
+
+  @override
+  void discard() {
+    super.discard();
+    // Do not detach the child container from the root. It is permanently
+    // attached. The elements are reused together and are detached from the DOM
+    // together.
+    _childContainer = null;
+  }
+
+  @override
+  html.Element createElement() {
+    html.Element element = defaultCreateElement('flt-shader-mask');
+    html.Element container = html.Element.tag('flt-mask-interior');
+    container.style.position = 'absolute';
+    _childContainer = container;
+    element.append(_childContainer!);
+    return element;
+  }
+
+  @override
+  void apply() {
+    if (_shaderElement != null) {
+      _shaderElement?.remove();
+    }
+
+    // TODO(flutter-web): Support non-gradient shaders for ShaderMask.
+    if (!(shader is ui.Gradient)) {
+      return;
+    }
+    final gradient = shader as ui.Gradient;
+    // TODO(flutter-web): Support conical and sweep gradients for ShaderMask.
+    if (gradient is GradientConical || gradient is GradientSweep) {
+      return;
+    }
+
+    String inputImageSvg;
+    if (gradient is GradientLinear) {
```

Comment:
would be nice to refactor this into a utility function :  _gradientToSvgElement(EngineGradient gradient...) , so in the error case you can throw ArgumentError.